### PR TITLE
abstracting getting variation product id from variation selections

### PR DESCRIPTION
**Pull request by curtismchale**
Here we pull out the loops in wpsc_add_to_cart that took the variation
selection and parsed it down to a variation_product_id and moved them in
to their own function.

This allows developers to leverage a core WPSC function instead of
maintaining their own loops to get the variation_product_id and then
needing to maintain them long term.

It also breaks out the huge function that is wpsc_add_to_cart in to one
more smaller part that we could write a unit test for without it being
buried deep down inside another function.

See #1813 for initial issue submission.

- this is tagged as @since 4.0 so that shoud be verified before we merge this

### CODE REVIEW

File: wpsc-components/theme-engine-v1/helpers/ajax.php [hunk 1]
```diff
@@ -84,6 +84,8 @@ function wpsc_special_widget() {
 /**
  * add_to_cart function, used through ajax and in normal page loading.
  * No parameters, returns nothing
+ *
+ * @uses wpsc_get_product_id_from_variation_selections()                Given array of variation selections returns the variation product id as int
```
Comment (JustinSainton) [inline wpsc-components/theme-engine-v1/helpers/ajax.php]: The long description here needs to be on its own line.

File: wpsc-includes/variations.class.php [hunk 2]
```diff
@@ -296,3 +296,30 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
 		return false;
 	}
 }
+
+/**
+ * Given array of variation selections this works through the terms and returns the product_id for the matching variation
+ *
+ * @since 4.0
+ * @author Curtis McHale
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Our doc standards don't include adding an author parameter.

File: wpsc-includes/variations.class.php [hunk 3]
```diff
@@ -296,3 +296,30 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+ *
+ * @param array     $variations     required            The array of variation selections
+ * @uses wpsc_get_child_objects_in_terms()              Given $product_id and product params this returns the variation product id
+ * @return int      $product_id                         The ID of the variation product
+ */
+function wpsc_get_product_id_from_variation_selections( $variations ){
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Let's call this `wpsc_get_product_id_from_variations`.  

Also, need a space between the bracket and closing parentheses.

File: wpsc-includes/variations.class.php [hunk 4]
```diff
@@ -296,3 +296,30 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+
+	foreach ( (array) $variations['variation'] as $key => $variation ) {
+		$provided_parameters['variation_values'][ (int) $key ] = (int) $variation;
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: `$provided_parameters` is not a defined variable.

File: wpsc-includes/variations.class.php [hunk 5]
```diff
@@ -296,3 +296,30 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+	}
+
+	if ( count( $provided_parameters['variation_values'] ) > 0 ) {
+		$variation_product_id = wpsc_get_child_object_in_terms( $product_id, $provided_parameters['variation_values'], 'wpsc-variation' );
+		if ( $variation_product_id > 0 ) {
+			$product_id = $variation_product_id;
+		}
+	}
+
+	return absint( $product_id );
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Let's filter the return value as well.

Comment (curtismchale) [inline wpsc-includes/variations.class.php]: Clarify please? You mean just remove the absint or ...?

File: wpsc-components/theme-engine-v1/helpers/ajax.php (hunk 1 above)
Comment (curtismchale) [inline wpsc-components/theme-engine-v1/helpers/ajax.php]: Sorry not following.

File: wpsc-includes/variations.class.php (hunk 5 above)
Comment (curtismchale) [inline wpsc-includes/variations.class.php]: bah missed the word 'filter' nevermind.

File: wpsc-includes/variations.class.php [hunk 6]
```diff
@@ -296,3 +296,39 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+function wpsc_get_product_id_from_variations( $variations ) {
+
+	$provided_parameters = array();
+
+	foreach ( (array) $variations['variation'] as $key => $variation ) {
+		$provided_parameters['variation_values'][ (int) $key ] = (int) $variation;
+	}
+
+	if ( count( $provided_parameters['variation_values'] ) > 0 ) {
+		$variation_product_id = wpsc_get_child_object_in_terms( $product_id, $provided_parameters['variation_values'], 'wpsc-variation' );
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: `$product_id` is not defined.

Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Also - I have always sort of hated the name `$provided_parameters`.  Let's name this something...anything...better.  Your call.

Comment (curtismchale) [inline wpsc-includes/variations.class.php]: I changed the variable name here but not in the main wpsc_add_to_cart function at this point. In my mind changing it there would be it's own PR because it has a bunch of changes to me made that could break things and it wouldn't affect whether this works or not.

File: wpsc-includes/variations.class.php [hunk 7]
```diff
@@ -296,3 +296,50 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
 		return false;
 	}
 }
+
+/**
+ * Given array of variation selections this works through the terms and returns the product_id for the matching variation
+ *
+ * @since 4.0
+ *
+ * @param array     $variations     required            The array of variation selections
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Let's change this function to `wpsc_get_product_data_from_variations`.

Then, let's have two wrapper functions for it, `wpsc_get_product_id_from_variations` and `wpsc_get_variation_values_from_variations`.

Finally, let's document clearly the expectations for the `$variations` parameter.  If the goal is to have this function abstracted to be useful elsewhere, we need to clearly describe the array of data intended to be passed.

File: wpsc-includes/variations.class.php [hunk 8]
```diff
@@ -296,3 +296,117 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+
+	/**
+	 * Allows users to filter the product_id based on the variation selections
+	 *
+	 * @since 4.0
+	 *
+	 * @param int   $product_id             The default passed product_id
+	 * @param array $variations             The variation selections passed to the core function
+	 */
+	$product_id = apply_filters( 'wpsc_variation_product_id', absint( $values['product_id'] ), $variations );
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Let's add `$values` and the originally passed `$product_id` as additional parameters to the filter.

And instead of setting the $product_id variable here, let's simply return it, e.g. `return absint( apply_filters( 'wpsc_variation_product_id', absint( $values['product_id'] ), $variations, $product_id, $values ) );`

File: wpsc-includes/variations.class.php [hunk 9]
```diff
@@ -296,3 +296,117 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+
+	/**
+	 * Allows users to filter the product_id based on the variation selections
+	 *
+	 * @since 4.0
+	 *
+	 * @param int   $product_id             The default passed product_id
+	 * @param array $variations             The variation selections passed to the core function
+	 */
+	$product_id = apply_filters( 'wpsc_variation_product_id', absint( $product_id ), $variations );
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: I like that the product ID is filtered the same way in both functions, but rather than duplicating the filter, let's do it once, and reference it in the other, like [WordPress does](https://github.com/WordPress/WordPress/blob/master/wp-includes/option.php#L99-L100).

Also, let's make sure that we're able to pass the original `$product_id` passed into the function, as well as the newly retrieved `$variation_product_id`.

File: wpsc-includes/variations.class.php [hunk 10]
```diff
@@ -296,3 +296,120 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+/**
+ * Given array of variation selections this works through the terms and returns the product_id for the matching variation
+ *
+ * For the `$variations` parameter we expect to get two 'term_ids' which correspond to the selections in the variation.
+ * So if we have a blue large shirt and the blue term_id is 8 and the large term_id is 12 we would get an array that looks like.
+ *
+ * array(
+ *      '2' => '8',     // this is the blue term_id
+ *      '9' => '12',    // this is the large term_id
+ * );
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: This example doesn't clearly communicate what the '2' and '9' are.

File: wpsc-includes/variations.class.php [hunk 11]
```diff
@@ -296,3 +296,120 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+ *
+ * Here the keys are captured when someone clicks the 'add to cart' button and correspond with ... whatever. They don't really matter for our function.
+ * Really ou could pass an array of 2 term_ids here and a $product_id that has variations to match those terms and you'd get back the expected array.
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: s/ou/you

File: wpsc-includes/variations.class.php [hunk 12]
```diff
@@ -296,3 +296,120 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+ *
+ * @since 4.0
+ *
+ * @param array     $variations     required            The array of variation selections
+ * @param int       $product_id     required            The default product_id
+ * @uses wpsc_get_child_objects_in_terms()              Given $product_id and product params this returns the variation product id
+ * @return array    $args {
+ *      @param  int     product_id                      The variation product_id
+ *      @param  array   variation_values                The array of variation_values that wpsc_add_to_cart needs to populate $provided_parameters['variation_values']
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Spacing here is odd.

File: wpsc-includes/variations.class.php [hunk 13]
```diff
@@ -296,3 +296,120 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+	 * Allows users to filter the product_id based on the variation selections
+	 *
+	 * @since 4.0
+	 *
+	 * @param int   $$values['product_id']  The variation product_id
+	 * @param array $variations             The variation selections passed to the core function
+	 * @param int   $product_id             The originally passed $product_id
+	 * @param array $values{
+	 *      @param  int     product_id         The variation product_id
+	 *      @param  array   variation_values   The array of variation term_ids based on the selections sent through
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Spacing here is odd.

File: wpsc-includes/variations.class.php [hunk 14]
```diff
@@ -296,3 +296,120 @@ function wpsc_get_child_object_in_terms_var( $parent_id, $terms, $taxonomies, $a
...
+	/**
+	 * Allows users to filter the variation values based on the variation selections
+	 *
+	 * @since 4.0
+	 *
+	 * @param array $variations             The variation selections passed to the core function
+	 * @param int   $product_id             The default passed product_id
+	 * @param array $values{
+	 *      @param  int     product_id         The variation product_id
+	 *      @param  array   variation_values   The array of variation term_ids based on the selections sent through
```
Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Odd spacing here as well.

Comment (curtismchale) [inline wpsc-includes/variations.class.php]: can you clarify what's odd about it? Should be tabs to start a line and spaces in the middle of the line. That's what my editor (PHPStorm) shows me.

Based on the WordPress doc standards I have the returns in the correct place as far as I can tell as well.

Comment (JustinSainton) [inline wpsc-includes/variations.class.php]: Derp, don't mind me, didn't realize those docs were for the keys in the array.

That said, there are still minor spacing issues - MUCH more minor, like a space after `array $values` and before the closing bracket.

https://make.wordpress.org/core/handbook/inline-documentation-standards/php-documentation-standards/#1-1-parameters-that-are-arrays
